src/replica/testHttpAsyncReq.cc

```diff
@@ -0,0 +1,546 @@
+    // being stuck for longer than expected.
+    ::AsyncTimer testAbortTimer(io_service, 100, [] (auto expirationIvalMs) {
+        LOGS_INFO("HttpAsyncReq_body_limit_error: test exceeded the time budget of " << expirationIvalMs << " ms");
+        std::exit(1);
+    });
+    testAbortTimer.start();
+
+    // Set up and start the server
+    ::HttpServer httpServer;
+    size_t const serverResponseBodySize = 1024;
+    httpServer.server()->addHandler("PUT", "/return_large_body",
+        [serverResponseBodySize] (qhttp::Request::Ptr const& req, qhttp::Response::Ptr const& resp) {
+            string const largeBody(serverResponseBodySize, 'a');
+            string const contentType = "text/html";
+            resp->send(largeBody, contentType);
+        }
+    );
+    httpServer.start();
+
+    // Submit a request.
+    string const url = "http://127.0.0.1:" + to_string(httpServer.port()) + "/return_large_body";
+    string const method = "PUT";
+    string const data;
+    unordered_map<string, string> const headers;
+    size_t const maxResponseBodySize = serverResponseBodySize - 1;
+    shared_ptr<HttpAsyncReq> const req = HttpAsyncReq::create(
+        io_service,
+        [&testAbortTimer, serverResponseBodySize](auto const& req) {
+            testAbortTimer.cancel();
+            BOOST_CHECK(req->state() == HttpAsyncReq::State::BODY_LIMIT_ERROR);
+            BOOST_CHECK(req->errorMessage().empty());
+            BOOST_CHECK_EQUAL(req->responseCode(), 200);
+            BOOST_CHECK_EQUAL(req->responseHeader().at("Content-Length"), to_string(serverResponseBodySize));
+            BOOST_CHECK_EQUAL(req->responseHeader().at("Content-Type"), "text/html");
+            BOOST_CHECK_THROW(req->responseBody(), std::logic_error);
+            BOOST_CHECK_THROW(req->responseBodySize(), std::logic_error);
+        },
+        method, url, data, headers, maxResponseBodySize
+    );
+    req->start();
+
+    thread serviceThread([&io_service] () {
+        io_service.run();
+    });
+    serviceThread.join();
+}
+
+// Testing request expiration due to non-responsive server (which is simulated
+// by introducing a delay into the request handler.)
+
+BOOST_AUTO_TEST_CASE(HttpAsyncReq_expired) {
+
+    LOGS_INFO("HttpAsyncReq_expired");
+
+    asio::io_service io_service;
+
+    // The deadline timer limits the duration of the test to prevent the test from
+    // being stuck for longer than expected.
+    ::AsyncTimer testAbortTimer(io_service, 3000, [] (auto expirationIvalMs) {
+        LOGS_INFO("HttpAsyncReq_expired: test exceeded the time budget of " << expirationIvalMs << " ms");
+        std::exit(1);
+    });
+    testAbortTimer.start();
+
+    // Set up and start the server
+    ::HttpServer httpServer;
+    httpServer.server()->addHandler("POST", "/delayed_response",
+        [] (qhttp::Request::Ptr const& req, qhttp::Response::Ptr const& resp) {
+            this_thread::sleep_for(chrono::milliseconds(2500));
```

Comment:
This `sleep` is probably jamming the server (which has one thread) from processing further requests (and if you had run the entire test on a single thread would jam the entire test).  Since you are in an asio thread here, instead of sleep, it would probably be better to use an async timer (capture the response pointer in the completion lambda)

Reply:
Yes, I'm aware of this risk. Fortunatelly, this is not a problem here because `qhttp` has its own instance of `io_service` and a dedicated thread. Besides, each of the test cases initiates no more than one request.